### PR TITLE
(feat) fast append for new node

### PR DESCRIPTION
**Pull request by YoniFeng**
I presume that building the tree and only then (possibly) mutating/transforming it, is the most common usage pattern (baseless presumption, but also confirmed for existing dependents like `pact-reference`).
The current `Node.Id.append()` API isn't optimized for building up a tree from the ground up, where inserted nodes are new (therefore isolated/detached) and many of the checks and operations performed are redundant.

I'd like to suggest adding an API to allow for quicker building up of the tree.
Would love to hear your thoughts.

**Note**
The initial commits are very crude. Would like to get feedback before proceeding.
I think a cleaner API for users might be some sort of `append_value`, where the `T` value is supplied as an argument. Less verbose and hides the complexity (required assumptions).

### CODE REVIEW

**Comment by Boshen** [comment]
Background: The use case for [my project](https://github.com/Boshen/oxc) is to build up a parent pointing tree in the semantic analyzer:

https://github.com/Boshen/oxc/blob/e71de0bff607416cb1f9f42931f77dbd5e9a84ec/crates/oxc_semantic/src/builder.rs#L36-L46

And after some profiling, we discovered [some performance issues](https://github.com/Boshen/oxc/issues/115#issuecomment-1454692843) so hence this PR.

**Comment by YoniFeng** [comment]
I benchmarked locally with the latest 'fix' commit:

group                      main                                   pr
-----                      ----                                   --
semantic/babylon.max.js    2.18   178.1±42.13ms    58.0 MB/sec    1.00     81.8±1.74ms   126.2 MB/sec
semantic/d3.js             1.47     10.9±0.12ms    50.1 MB/sec    1.00      7.4±0.19ms    73.8 MB/sec
semantic/lodash.js         2.59      2.0±0.02ms   256.2 MB/sec    1.00   774.0±19.45µs   664.1 MB/sec
semantic/pdf.js            2.85      6.6±0.07ms    60.7 MB/sec    1.00      2.3±0.08ms   173.2 MB/sec
semantic/typescript.js     1.51     86.1±1.00ms   111.7 MB/sec    1.00     57.1±1.49ms   168.5 MB/sec

**Comment by YoniFeng** [comment]
@saschagrunert gentle ping

Distilled questions:
1. Thoughts on adding a fast path for a new node? (yay/nay)
2. Do you have a preference between adding `pub fn append_new_node<T>(self, new_child: NodeId, arena: &mut Arena<T>)` (current commit/draft) or something like `pub fn append_new_value<T>(self, new_value: T, arena: &mut Arena<T>)` to the exposed API?

**Comment by saschagrunert** [comment]
Hey @YoniFeng, thank you for the PR! I generally think that it would be useful to extend the API to save some lines in user code. I'm happy with `append_new_value` (or `append_value`). Do you think we still need `append_new_node` here?

**Comment by YoniFeng** [comment]
Only one is needed, I commit just to have both visible.
I'll rename to `append_value` and remove `append_new_node`.

File: src/id.rs [hunk 1]
```diff
@@ -679,6 +681,67 @@ impl NodeId {
...
+    /// let mut arena = Arena::new();
+    /// let n1 = arena.new_node("1");
+    /// assert!(n1.checked_append(n1, &mut arena).is_err());
+    ///
+    /// let n1_1 = arena.new_node("1_1");
+    /// assert!(n1.checked_append(n1_1, &mut arena).is_ok());
+    /// ```
+    ///
+    /// [`remove`]: struct.NodeId.html#method.remove
+    fn unchecked_append_new_node<T>(self, new_child: NodeId, arena: &mut Arena<T>) {
```
Comment (YoniFeng) [inline src/id.rs:741]: nits: `unchecked` at beginning or end? need consistency

Comment (Boshen) [reply src/id.rs:741]: end of the function, by Rust conventions.

**Comment by YoniFeng** [comment]
PTAL / ready for submit AFAICT

Do I need to squash commits?

**Review by saschagrunert — APPROVED** [review]

**Comment by saschagrunert** [comment]
LGTM, thank you!

**Pull request merged**